include/flatbuffers/flatbuffers.h

```diff
@@ -542,57 +563,228 @@ struct String : public Vector<char> {
+      size_(other.size_),
+      cur_(other.cur_) {
+    assert(allocator_);
+    other.allocator_ = nullptr;
+    other.buf_ = nullptr;
+    other.size_ = 0;
+    other.cur_ = nullptr;
+  }
+
+  ~DetachedBuffer() {
+    reset(nullptr);
+  }
+
+  inline uint8_t *release() FLATBUFFERS_NOEXCEPT {
+    uint8_t *retval = buf_;
+    if (allocator_ != nullptr) {
+      delete allocator_;
+    }
+    allocator_ = nullptr;
+    buf_ = nullptr;
+    size_ = 0;
+    cur_ = nullptr;
+    return retval;
+  }
+
+  inline void reset(std::nullptr_t /* unused */ = nullptr) FLATBUFFERS_NOEXCEPT {
+    if (buf_ != nullptr) {
+      assert(allocator_ != nullptr);
+      allocator_->deallocate(buf_, size_);
+    }
+    if (allocator_ != nullptr) {
+      delete allocator_;
+    }
+    allocator_ = nullptr;
+    buf_ = nullptr;
+    size_ = 0;
+    cur_ = nullptr;
+  }
+
+  inline uint8_t *get() const FLATBUFFERS_NOEXCEPT {
+    return cur_;
+  }
+
+  inline uint8_t operator*() const {
+    return *cur_;
+  }
+
+  inline uint8_t *operator->() const FLATBUFFERS_NOEXCEPT {
+    return cur_;
+  }
+
+  // These may change access mode, leave these at end of public section
+  FLATBUFFERS_DELETE_FUNC(DetachedBuffer(const DetachedBuffer &other))
+  FLATBUFFERS_DELETE_FUNC(DetachedBuffer &operator=(const DetachedBuffer &other))
+
+ protected:
+  Allocator* allocator_;
+  uint8_t *buf_;
+  size_t size_;
+  uint8_t *cur_;
 };
 
+// Add deprecated typedef for legacy code, to be removed in a future version.
+DEPRECATED("use flatbuffers::DetachedBuffer instead")
+typedef DetachedBuffer unique_ptr_t;
+
 // This is a minimal replication of std::vector<uint8_t> functionality,
 // except growing from higher to lower addresses. i.e push_back() inserts data
 // in the lowest address in the vector.
 class vector_downward {
  public:
   explicit vector_downward(size_t initial_size,
-                           const simple_allocator &allocator)
-    : reserved_((initial_size + sizeof(largest_scalar_t) - 1) &
-        ~(sizeof(largest_scalar_t) - 1)),
-      buf_(allocator.allocate(reserved_)),
-      cur_(buf_ + reserved_),
-      allocator_(allocator) {}
+                           Allocator *allocator,
+                           bool steal_allocator)
+    : allocator_(steal_allocator ? allocator : new AllocatorProxy(allocator)),
+      reserved_((initial_size + sizeof(largest_scalar_t) - 1) &
+                ~(sizeof(largest_scalar_t) - 1)),
+      buf_(allocator_->allocate(reserved_)),
+      cur_(buf_ + reserved_) {
+    assert(allocator_);
+  }
 
   ~vector_downward() {
-    if (buf_)
-      allocator_.deallocate(buf_);
+    if (buf_ != nullptr) {
+      assert(allocator_ != nullptr);
+      allocator_->deallocate(buf_, reserved_);
+    }
+    if (allocator_ != nullptr) {
+      delete allocator_;
+    }
   }
 
-  void clear() {
-    if (buf_ == nullptr)
-      buf_ = allocator_.allocate(reserved_);
+  void reset(size_t initial_size, Allocator *allocator, bool steal_allocator) {
```

Comment:
It's used in the other gRPC-related PR. When a message is popped out of a builder, we want to reset the vector with an allocator so it remains valid. Otherwise as soon as we pop out the message, the vector/builder is a landmine with nullptrs inside of it. I can take this out and rework that if necessary.